src/platform/silabs/tracing/SilabsTracing.h

```diff
@@ -0,0 +1,205 @@
+
+    // This Either Logs the traces or stores them in a buffer
+    CHIP_ERROR OutputTimeTracker(const TimeTracker & tracker);
+    CHIP_ERROR OutputWaterMark(TimeTraceOperation aOperation);
+    CHIP_ERROR OutputAllWaterMarks();
+    CHIP_ERROR TraceBufferFlushAll();
+    CHIP_ERROR TraceBufferFlushByOperation(TimeTraceOperation aOperation);
+
+    // prevent copy constructor and assignment operator
+    SilabsTracer(SilabsTracer const &)             = delete;
+    SilabsTracer & operator=(SilabsTracer const &) = delete;
+
+    // Methods to get the time trackers metrics values
+    TimeTracker GetTimeTracker(TimeTraceOperation aOperation) { return mLatestTimeTrackers[to_underlying(aOperation)]; }
+    Watermark GetWatermark(TimeTraceOperation aOperation) { return mWatermarks[to_underlying(aOperation)]; }
+
+    // Method to save the time trackers in the NVM, this will likely be time consuming and should not be called frequently
+    CHIP_ERROR SaveWatermarks();
+
+    // Method to load the time trackers from the NVM, this will be called after the initialization so the bootup watermarks need to
+    // be added to the loaded values
+    CHIP_ERROR LoadWatermarks();
+
+#if CONFIG_BUILD_FOR_HOST_UNIT_TEST
+    CHIP_ERROR GetTraceCount(size_t & count) const;
+    CHIP_ERROR GetTraceByIndex(size_t index, const char *& trace) const;
+#endif
+
+private:
+    struct TimeTrackerList
+    {
+        chip::SingleLinkedListNode<TimeTracker> * head = nullptr;
+
+        void Insert(const TimeTracker & tracker)
+        {
+            auto * newNode =
+                static_cast<chip::SingleLinkedListNode<TimeTracker> *>(malloc(sizeof(chip::SingleLinkedListNode<TimeTracker>)));
+            if (newNode != nullptr)
+            {
+                newNode->mValue = tracker;
+                newNode->mpNext = head;
+                head            = newNode;
+            }
+        }
+
+        void Remove(size_t index)
+        {
+            if (index == 0 && head != nullptr)
+            {
+                auto * temp = head;
+                head        = head->mpNext;
+                free(temp);
+                return;
+            }
+
+            auto * current = head;
+            for (size_t i = 0; i < index - 1 && current != nullptr; ++i)
+            {
+                current = current->mpNext;
+            }
+
+            if (current && current->mpNext)
+            {
+                auto * temp     = current->mpNext;
+                current->mpNext = current->mpNext->mpNext;
+                free(temp);
+            }
+        }
+    };
+
+    // Singleton class with a static instance
+    static SilabsTracer sInstance;
+
+    SilabsTracer();
+
+    // List of past time trackers
+    TimeTrackerList mTimeTrackerList;
+
+    // Time trackers to store time stamps for ongoing operations
+    TimeTracker mLatestTimeTrackers[kNumTraces];
+
+    // Watermarks for each operation
+    Watermark mWatermarks[kNumTraces];
```

Comment:
Parallel should be avoided if possible. Unless I missed something, these watermarks may be part of the TimeTracker structure.

Reply:
They used to be, then they got separated. They are not necessarily tied to each other as the TimeTrackers are only carriers of the latest changes whereas Watermarks will persist over several operations of the same kind.